cpu/stm32/periph/ptp.c

```diff
@@ -0,0 +1,216 @@
+     * register whenever a 32 bit accumulator register overflows*/
+    ETH->PTPSSIR = ptpssir;
+    ptp_clock_adjust_speed(0);
+    /* Wait new PTPSAR value becomes active */
+    while (ETH->PTPTSCR & ETH_PTPTSCR_TSARU) { }
+    /* Enable fine grained correction now */
+    ETH->PTPTSCR |= ETH_PTPTSCR_TSFCU;
+
+    static const ptp_timestamp_t initial_time = {
+        .seconds = 0,
+        .nanoseconds = 0
+    };
+    ptp_clock_set(&initial_time);
+    if (IS_USED(MODULE_PERIPH_ETH)) {
+        /* enable timestamping of all received frames */
+        ETH->PTPTSCR |= ETH_PTPTSSR_TSSARFE;
+    }
+    DEBUG("[periph_ptp] Initialized with PTPSAR = %" PRIu32 ", PTPSSIR = %" PRIu32 "\n",
+          ptptsar, ptpssir);
+}
+
+void ptp_clock_adjust_speed(int16_t correction)
+{
+    uint64_t offset = ptptsar;
+    offset *= correction;
+    offset >>= 16;
+    uint32_t adjusted_ptptsar = ptptsar + (uint32_t)offset;
+    /* Value to add onto the 32 bit accumulator register (which causes the
+     * value in ETH->PTPSSIR to be added onto the subsection register on 
+     * overflow) */
+    ETH->PTPTSAR = adjusted_ptptsar;
+    /* Wait for pending clock speed adjustments to complete */
+    while (ETH->PTPTSCR & ETH_PTPTSCR_TSARU) { }
+    /* Load new PTPTSAR value to hardware */
+    ETH->PTPTSCR |= ETH_PTPTSCR_TSARU;
+    DEBUG("[periph_ptp] Using PTPSAR = %" PRIu32 ", PTPSSIR = %" PRIu32 "\n",
+          adjusted_ptptsar, ptpssir);
+}
+
+void ptp_clock_adjust(int64_t offset)
+{
+    unsigned state = irq_disable();
+    ptp_timestamp_t ts;
+    uint64_t abs_offset = (offset < 0) ? -offset : offset;
+    ptp_ns2ts(&ts, abs_offset);
+    ETH->PTPTSHUR = ts.seconds;
+    ETH->PTPTSLUR = (offset < 0) ? (1UL << 31) | ts.nanoseconds : ts.nanoseconds;
+    while (ETH->PTPTSCR & (ETH_PTPTSCR_TSSTU | ETH_PTPTSCR_TSSTI)) {
+        /* wait until new time value can be set */
+    }
+    ETH->PTPTSCR |= ETH_PTPTSCR_TSSTU;
+    irq_restore(state);
+    DEBUG("[periph_ptp] Updated time by %c%" PRIu32 ".%09" PRIu32 "\n",
+          (offset < 0) ? '-' : '+', (uint32_t)ts.seconds, ts.nanoseconds);
+}
+
+void ptp_clock_set(const ptp_timestamp_t *time)
+{
+    assert(time && time->nanoseconds < NS_PER_SEC);
+    unsigned state = irq_disable();
+    /* First, set the timestamp update registers */
+    ETH->PTPTSHUR = time->seconds;
+    ETH->PTPTSLUR = time->nanoseconds;
+    /* From the data sheet (regarding setting TSSTI):
+     * > Both the TSSTU and TSSTI bits must be read as zero before you can set
+     * > this bit.
+     */
+    while (ETH->PTPTSCR & (ETH_PTPTSCR_TSSTU | ETH_PTPTSCR_TSSTI)) {
+        /* wait until new time value can be set */
+    }
+    /* Now, ask the peripheral to atomically set the clock from the update
+     * registers */
+    ETH->PTPTSCR |= ETH_PTPTSCR_TSSTI;
+    irq_restore(state);
+}
+
+void ptp_clock_read(ptp_timestamp_t *timestamp)
+{
+    /* Read first high register, then low, then again high. If the value in
```

Comment:
should this  be done with interrupts disabled?

Reply:
Should work fine without. But the lengthy comment above could be dropped if IRQs were just disabled.

Reply:
I changed my mind. The `irq_disable()` and `irq_restore()` functions also contain barriers.

E.g. doing

```C
ptpt_timestamp_t pre, post;
ptp_clock_read(&pre);
do_some_cpu_intensive_work();
ptp_clock_read(&post);
```

Should not be optimized by the compiler to
```C
ptpt_timestamp_t pre, post;
ptp_clock_read(&pre);
ptp_clock_read(&post);
do_some_cpu_intensive_work();
```

So I really want the barriers there to prevent above optimization to not introduce bugs only triggered with `LTO=1`.